### PR TITLE
fix(proxycache): retry interrupted upstream blob fetches, verify digest before caching

### LINKED ISSUE #23600: Proxy cache: no retry on interrupted upstream blob fetch, and no digest verification before caching
**Expected behavior and actual behavior:**

Expected: when Harbor's proxy cache is pulling a blob (layer) from an upstream registry and the upstream connection is interrupted mid-transfer (dropped/reset by an intermediate network device), Harbor should retry the fetch — comparable to how a plain docker/OCI client retries an interrupted layer pull — and should never cache or serve a blob whose content doesn't match its declared digest.

Actual: the proxy cache's upstream blob fetch is a single-shot HTTP GET with no retry logic anywhere in the code path, and no digest verification of the fetched bytes before they're pushed to local cache storage. Concretely:

- `ProxyBlob` / `putBlobToLocal` in `src/controller/proxy/controller.go` each call `RemoteInterface.BlobReader()` exactly once. On a mid-stream read error (e.g. `unexpected EOF` from a reset connection), the client-serving fetch just returns the error, and the background cache-populate fetch logs `"error while putting blob to local repo, %v"` and gives up — no retry, no resume from the last byte offset.
- `ManifestCache.putBlobToLocal` in `src/controller/proxy/manifestcache.go` has the identical single-shot, no-retry pattern, used when caching a manifest list's dependent blobs.
- Neither path recomputes/verifies the digest of the bytes actually received against the expected blob digest before treating a cache-populate push as successful — verification is implicitly delegated to whatever the local registry storage engine happens to do at commit time, not performed by Harbor itself.

In our deployment, images are pulled through Harbor's proxy cache in front of an upstream registry, with a corporate TLS-inspecting web proxy sitting between Harbor and the upstream (CRI-O → Harbor proxy cache → corporate web proxy → upstream registry). That web proxy has a non-trivial false-positive rate on its inspection rules and, when triggered on a specific blob, kills the TCP session mid-transfer. Because Harbor doesn't retry, this either fails the cache-populate fetch outright (visible in logs) or — depending on exactly where in the stream the reset lands relative to header/body framing — can result in an incomplete/corrupted blob reaching the cache-populate path with nothing catching the mismatch before it's pushed. Our current workaround is pulling the same image with the plain `docker` CLI through the same corporate proxy instead of through Harbor, since the docker/containerd client-side pull logic retries the dropped connection and the proxy doesn't re-trigger on the retry — this strongly suggests the fix belongs on Harbor's proxy-cache fetch path, not the network layer.

**Steps to reproduce the problem:**

1. Configure a Harbor proxy-cache project pointing at an upstream registry that is reachable only through a network path capable of resetting a TCP connection mid-transfer (a corporate TLS-inspecting proxy, a flaky link, or any MITM/inspection proxy that can be induced to drop a session mid-stream).
2. Pull a multi-layer image through the proxy-cache project (`docker pull <harbor-host>/<proxy-project>/<image>:<tag>`), ideally one with at least one large layer, so there's a realistic window for a mid-transfer interruption.
3. Trigger (or wait for) the intermediate network path to reset the connection while Harbor is mid-fetch of one of the blobs from upstream.
4. Observe harbor-core logs for the proxy-cache background push, e.g.:
   ```
   error while putting blob to local repo, <error containing "unexpected EOF" or a similar mid-stream read error>
   ```
   at `src/controller/proxy/controller.go`'s `putBlobToLocal` (or the equivalent path in `manifestcache.go` if the interrupted blob is a manifest-list dependency).
5. Note that nothing retries this fetch — the layer either fails to get cached (subsequent pulls repeat the same failure-prone fetch against the same flaky network path) or, if the interruption manifests as a length-preserving corruption rather than a hard read error, nothing in Harbor's own code checks the resulting bytes against the blob's digest before caching.

**Versions:**
Please specify the versions of following systems.

- harbor version: `v2.15.2`
- docker engine version: not applicable — this is a proxy-cache upstream-fetch code path, reproducible independent of the pulling client's docker engine version
- docker-compose version: not applicable

**Additional context:**

- Relevant code: `src/controller/proxy/controller.go` (`ProxyBlob`, `putBlobToLocal`) and `src/controller/proxy/manifestcache.go` (`ManifestCache.putBlobToLocal`) — both call `RemoteInterface.BlobReader()` a single time with no retry/backoff, and neither verifies the fetched content's digest before treating the fetch as complete/cacheable.
- This looks related to, but distinct from, a few existing reports, which I think are worth cross-referencing rather than treating as duplicates:
  - #20831 ("Image Digest Mismatch Issue") — the reporter's logs show exactly this code path failing (`error while putting blob to local repo` at `controller.go`, with the underlying registry logging `client disconnected during blob PUT ... unexpected EOF`) on a large blob push to local cache. The response there addressed a narrower, separate digest-mismatch cause, not the underlying lack of retry.
  - #22393 ("Harbor image pull fails for large images (~40GB) after ~30 minutes despite timeout config changes") — root-caused to the registry HTTP client's overall request timeout (`REGISTRY_HTTP_CLIENT_TIMEOUT`, default 30 min) firing, which is a *timeout*, not a retry mechanism; closed as not planned. Same underlying gap (no retry on a failed upstream fetch), different trigger (timeout vs. connection reset).
  - #22555 ("Large image filesystem layer verification failed for digest") — first-time pulls of large images through proxy cache fail with a layer verification error; retrying the pull once the blob is cached elsewhere works. Consistent with a fetch-race/no-retry gap rather than a digest-verification bug per se.
  - #14822 ("Proxy cache does not provide images when target registry is not working") — related reliability gap in the same area: the proxy cache has no resilience story for upstream connectivity problems in general, whether that's a hard outage (#14822) or a mid-transfer interruption (this issue).
- I have a fix ready and am opening a PR for it immediately alongside this issue: a bounded-retry reader that resumes an interrupted upstream fetch from the last successfully-read byte offset via an HTTP range request (PullBlobChunk, which already exists on the registry adapter but wasn't wired into the proxy-cache fetch path), plus a digest-verifying reader that rejects a cache-populate push whose content doesn't match the expected digest instead of silently committing it.

### PR DESCRIPTION
**Pull request by oscrx**
Thank you for contributing to Harbor!

# Comprehensive Summary of your change

The proxy cache's blob fetch made a single-shot HTTP GET to the upstream
registry with no retry: a connection dropped mid-stream (e.g. by a network
intermediary killing the TCP session, as described in #23600) just failed
the pull, with no equivalent of docker/containerd's client-side retry of an
interrupted layer. This affected both `controller.putBlobToLocal`
(background cache population after a client pull) and the identical
`ManifestCache.putBlobToLocal` used when caching a manifest list's
dependent blobs.

This adds a `resumingBlobReader` that, on a retryable mid-stream read error,
closes the broken connection and resumes the fetch from the last
successfully read byte offset via an HTTP range request
(`RemoteInterface.BlobReaderAt`, backed by the existing `PullBlobChunk`
which wasn't previously wired into the proxy-cache fetch path), with bounded
exponential backoff (3 attempts, 1s/2s/4s). It's applied to the
client-serving fetch and both cache-populate fetches.

It also adds a `verifyingReadCloser` that recomputes the digest of the
bytes streamed into each cache-populate path and surfaces a mismatch as a
Read error (instead of `io.EOF`), so a corrupted-but-well-formed body can't
be committed to local storage as if it were a complete, valid blob.

No config surface, API, or docs are affected — the retry count/backoff are
internal constants, not exposed as project/system configuration.

# Issue being fixed
Fixes #23600

Please indicate you've done the following:
- [x] Well Written Title and Summary of the PR
- [ ] Label the PR as needed. "release-note/ignore-for-release, release-note/new-feature, release-note/update, release-note/enhancement, release-note/community, release-note/breaking-change, release-note/docs, release-note/infra, release-note/deprecation"
- [x] Accepted the DCO. Commits without the DCO will delay acceptance.
- [x] Made sure tests are passing and test coverage is added if needed.
- [x] Considered the docs impact — no config/API surface changed, so no docs update is needed.

### CODE REVIEW

File: src/controller/proxy/remote.go [hunk 1]
```diff
@@ -100,6 +104,17 @@ func (r *remoteHelper) BlobReader(repo, dig string) (int64, io.ReadCloser, error
 	return sz, bReader, err
 }
 
+func (r *remoteHelper) BlobReaderAt(repo, dig string, size, start int64) (io.ReadCloser, error) {
+	_, bReader, err := r.registry.PullBlobChunk(repo, dig, size, start, size-1)
```
[the remark this answers is not in the document]
Comment (oscrx) [reply src/controller/proxy/remote.go:108]: Confirmed, thanks — verified by reading Go's net/http and PullBlobChunk directly (no test existed for this before). Fixed in 84e7c0db3: PullBlobChunk in src/pkg/registry/client.go now requires a 206 whose Content-Range actually starts at the requested offset, or a plain 200 only when start == 0; anything else returns an error instead of being treated as the chunk. Added TestPullBlobChunk_* covering the matching-206, honored-200-at-zero, ignored-range-at-nonzero, and mismatched-Content-Range cases.

File: src/controller/proxy/blobretry.go [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+		}
+		if !r.canRetry(err) {
+			return n, err
+		}
+		if rErr := r.reconnect(err); rErr != nil {
+			return n, rErr
+		}
+		if n > 0 {
+			return n, nil
+		}
```
[the remark this answers is not in the document]
Comment (oscrx) [reply src/controller/proxy/blobretry.go:122]: Confirmed — a clean EOF short of the declared size (e.g. a stream terminated gracefully rather than reset) previously bypassed retry entirely. Fixed in 84e7c0db3: resumingBlobReader.Read now converts a clean io.EOF into io.ErrUnexpectedEOF whenever offset < size, so it consumes the retry budget the same as a transport error. Added TestResumingBlobReader_RetriesOnPrematureCleanEOF.

File: src/controller/proxy/blobretry.go [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+	return &verifyingReadCloser{ReadCloser: reader, verifier: dig.Verifier(), dig: dig}
+}
+
+func (v *verifyingReadCloser) Read(p []byte) (int, error) {
+	n, err := v.ReadCloser.Read(p)
+	if n > 0 {
+		_, _ = v.verifier.Write(p[:n])
+	}
+	if err == io.EOF && !v.verifier.Verified() { // nolint:errorlint
+		return n, fmt.Errorf("blob content does not match expected digest %s after fetching from upstream", v.dig)
```
[the remark this answers is not in the document]
Comment (oscrx) [reply src/controller/proxy/blobretry.go]: Confirmed, and thanks for catching this — I re-checked net/http/transfer.go directly: writeBody does wrap the request body in io.LimitReader(body, ContentLength), so a well-behaved reader that returns its final bytes with a nil error (e.g. bytes.Reader, and plausibly the real response body) never gets a follow-up Read call to observe EOF, meaning Verified() could go unchecked. Fixed in 84e7c0db3: verifyingReadCloser now tracks the expected size and checks Verified() as soon as that many bytes have been read, in addition to the EOF fallback for when size is unknown. Added TestVerifyingReadCloser_VerifiesAssoonAsSizeReachedWithoutFurtherEOFCall, which reproduces the LimitReader-style single-call pattern with bytes.Reader.

File: src/pkg/registry/client.go [hunk 4]
```diff
@@ -398,6 +398,31 @@ func (c *client) PullBlobChunk(repository, digest string, _ int64, start, end in
...
+		got, crErr := parseContentRangeStart(cr)
+		if crErr != nil || got != start {
+			defer resp.Body.Close()
+			return 0, nil, fmt.Errorf("registry returned Content-Range %q for requested range bytes=%d-%d, refusing to treat it as that chunk", cr, start, end)
+		}
+	case http.StatusOK:
+		if start != 0 {
+			defer resp.Body.Close()
+			return 0, nil, fmt.Errorf("registry ignored range request bytes=%d-%d and returned a full 200 response", start, end)
+		}
```
[the remark this answers is not in the document]
Comment (oscrx) [reply src/pkg/registry/client.go]: Confirmed, thanks — I re-checked copyBlobByChunk directly and the first chunk request is exactly the case you describe (start==0, end < blobSize-1 for any blob bigger than one chunk). Fixed in c177a8d1b: PullBlobChunk now uses the blobSize parameter (previously discarded) to require start == 0 && end == blobSize-1 before accepting a 200, falling back to the weaker start==0-only check only when blobSize is unknown. Added TestPullBlobChunk_RejectsFull200ForPartialFirstChunk; also re-ran the full controller/replication/... suite to confirm this doesn't break the existing chunked-copy path.

File: src/controller/proxy/blobretry.go [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+// establishing the new connection also fails.
+func (r *resumingBlobReader) reconnect(cause error) error {
+	_ = r.reader.Close()
+	for {
+		r.retries++
+		backoff := blobFetchRetryBackoff * time.Duration(uint64(1)<<uint(r.retries-1))
+		log.Warningf("proxy cache: upstream blob read interrupted at byte %d/%d (%v), resuming attempt %d/%d in %s",
+			r.offset, r.size, cause, r.retries, maxBlobFetchRetries, backoff)
+		time.Sleep(backoff)
+		next, err := r.resume(r.offset)
```
[the remark this answers is not in the document]
Comment (oscrx) [reply src/controller/proxy/blobretry.go:148]: Confirmed — RemoteInterface/registry.Client don't accept a context anywhere, so an in-flight resumed request genuinely can't be aborted without a much larger interface change across the whole registry client, which felt out of scope here. What I could reasonably fix in this PR: resumingBlobReader now takes a context.Context and checks it before each backoff wait and before starting each further resume attempt, so a canceled caller stops the retry loop from sleeping out backoffs and opening new upstream requests once it's given up (it just can't cancel a request already in flight). The client-serving fetch in ProxyBlob is bound to the request context (canceled when the pulling client disconnects); both cache-populate fetches (controller.putBlobToLocal and ManifestCache.putBlobToLocal) intentionally use context.Background(), since that background caching work should keep going even after the client that triggered it is gone. Fixed in c177a8d1b, with TestResumingBlobReader_StopsRetryingWhenContextCanceled covering the cancellation path.

File: src/controller/proxy/blobretry_test.go [hunk 6]
```diff
@@ -0,0 +1,219 @@
...
+	content := []byte("corrupted blob content")
+	wrongDigest := digest.FromBytes([]byte("something else entirely"))
+
+	reader := newVerifyingReadCloser(io.NopCloser(&stubReader{data: content}), wrongDigest, int64(len(content)))
+	_, err := io.ReadAll(reader)
+	require.Error(t, err)
+	assert.Contains(t, err.Error(), "does not match expected digest")
+}
+
+func TestVerifyingReadCloser_VerifiesAssoonAsSizeReachedWithoutFurtherEOFCall(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (oscrx) [reply src/controller/proxy/blobretry_test.go]: Good catch, thanks — fixed in 4dafa9878.

File: src/controller/proxy/blobretry.go [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+	return &resumingBlobReader{ctx: ctx, reader: reader, size: size, resume: resume}
+}
+
+func (r *resumingBlobReader) Read(p []byte) (int, error) {
+	for {
+		n, err := r.reader.Read(p)
+		r.offset += int64(n)
+		if err == nil {
+			return n, err
+		}
```
[the remark this answers is not in the document]
Comment (oscrx) [reply src/controller/proxy/blobretry.go:101]: Confirmed — traced through it: once the final bytes bring offset to the full declared size, canRetry() was returning false purely because offset is no longer < size, and the accompanying error was propagated anyway even though the blob was actually complete. Fixed in 638cb24d9: the reader now checks for having reached the declared size before looking at the error at all, and treats that as success unconditionally. Note a naive "just return success once size is reached" fix would create a (0, nil) loop on any further call if the caller keeps reading (since the underlying reader may now be stale/exhausted) — so I added a completed flag that makes the reader self-terminate (io.EOF) on subsequent calls instead. Added TestResumingBlobReader_TreatsErrorAccompanyingFinalBytesAsSuccess with a oneShotReader that returns the final bytes and io.ErrUnexpectedEOF together in one call, matching the scenario you described.

**Comment by Vad1mo** [comment]
What negative effect might a retry have on rate limits? 
If a negative effect is possible, I would suggest to wrap this feature behind a feature flag.

**Comment by oscrx** [comment]
Good question. Two things worth separating here:

**How much extra load can retry actually add?** It only ever triggers on a genuinely *interrupted* connection (dropped/reset mid-transfer), never on a normal successful pull, and it's bounded to at most 3 extra attempts per interrupted blob with exponential backoff (1s/2s/4s) - not a multiplier on regular traffic. I'd actually argue this reduces aggregate upstream load in the failure case: today, an interrupted blob fails the whole pull outright, so the kubelet/docker client has to retry the *entire* image (manifest + every blob) from scratch. This resumes just the missing tail of the one interrupted blob via a Range request instead, which is strictly less upstream traffic than an outer full-image retry.

**Where you're right that there was a real gap:** if a *resume* attempt itself got rate-limited (429), the code was treating that the same as any other transient error and retrying it with the same 1-4s backoff - which won't help against a real rate limit (those typically need a much longer cool-down) and just adds more rejected requests. Fixed that in bbc3c780d: a rate-limited resume now gives up immediately instead of consuming the rest of the retry budget, using the existing errors.IsRateLimitError check (same one already used elsewhere in this file for manifest 429 handling).

Given the above I'd rather not add a feature flag - the behavior is already bounded and opt-out-shaped (only fires on failure), and a flag adds real config/test surface for a fairly narrow edge case. Happy to reconsider if you still think it's warranted, though.

**Comment by oscrx** [comment]
All the comments in the code and the commit messages are very verbose, I can condense a lot of that after initial review I think
